### DISCUSSION TITLE
Why a Workers Threads should load the WASM module into each Worker,  isn't it could be shared for all  Workers ?

### DISCUSSION THREAD

**Discussion by aharondavid** [emscripten-core/emscripten#20458] [Q&A]
Is seems that Worker Thread duplicate the WASM Module for every Thread, isn't it could be shared for all  Workers ? 
Is Wasm Worker (instead of Pthread Worker) is also load the WASM Module ?

**Comment by sbc100** [comment]
The wasm module is shared between threads vis postMessage.  But each thread needs to create its own instance of the wasm module.  Is that what you are referring to? 

Wasm instances cannot currently be shared between Web Workers.  This is limitation of the Web API.  Very few things in fact can be shared between workers.   As far as I know the only mutable thing that can be shared is SharedArrayBuffer.

**Comment by aharondavid** [comment]
Hi @sbc100 , look at the following steps that illustrate what I meant:

Here is a simple c code:
```
#include <emscripten.h>
#include <iostream>
#include <fstream>
float lerp(float a, float b, float t) {
    return (1 - t) * a + t * b;
}
int main(){}
```
	


let build it:
`emcc test.cpp  -s PTHREAD_POOL_SIZE=10 -s USE_PTHREADS   -s WASM=1 -o test.html`
insert the following script into test.html:
```
<script>    
    setInterval(function() {    
      performance.measureUserAgentSpecificMemory().then(function(memoryInfo) {
        console.log("JS memory usage: " + memoryInfo.bytes / 1024 / 1024 + " MB");
      });
    }, 2000);
  </script>
```
**test.wasm size is 37KB
test.js is 101KB**

Screenshot  of the Console result is attached, the memory allocated is about 180MB !
when I use  PTHREAD_POOL_SIZE=1 the size is 34MB 
when I remove the flag USE_PTHREADS  the size is 17MB
It seems that the WebAssembly.memory doesn't really shared.

Can you explain that behavior  ?

 
![image](https://github.com/emscripten-core/emscripten/assets/41375082/2e908f0c-9657-4e50-9a6f-ef40316e6cff)

**Reply by sbc100** [reply]
I believe what you are seeing here is the overhead of the WebWorker in which each thread needs to run.

The wasm memory is certainly shared, otherwise no threaded code would work (since threaded code fundamentally relies on shared memory).   If you want to prove that this is true you can try adding 100Mb to the size of the wasm memory (e.g. `-sINITIAL_MEMORY=200MB` then `-sINITIAL_MEMORY=300MB`).   In this case, regardless of how many threads you have running you should only see the overall memory usage go up by ~100Mb.

**Comment by aharondavid** [comment]
Yes, you right - the memory indeed shared, and it seems that nothing to be done with the size of the WebWorker.
See my conclusion about iOS Safari\Chrome here: https://github.com/emscripten-core/emscripten/discussions/20377
